src/shared/mount-util.c

```diff
@@ -1167,6 +1168,161 @@ int remount_idmap(
+        _cleanup_(mnt_free_iterp) struct libmnt_iter *iter = NULL;
+        SubMount *mounts = NULL;
+        size_t n = 0;
+        int r;
+
+        assert(prefix);
+        assert(ret_mounts);
+        assert(ret_n_mounts);
+
+        r = libmount_parse("/proc/self/mountinfo", NULL, &table, &iter);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to parse /proc/self/mountinfo: %m");
+
+        for (;;) {
+                _cleanup_close_ int mount_fd = -EBADF;
+                _cleanup_free_ char *p = NULL;
+                struct libmnt_fs *fs;
+                const char *path;
+
+                r = mnt_table_next_fs(table, iter, &fs);
+                if (r == 1)
+                        break;
+                if (r < 0) {
+                        log_debug_errno(r, "Failed to get next entry from /proc/self/mountinfo: %m");
+                        goto fail;
+                }
+
+                path = mnt_fs_get_target(fs);
+                if (!path)
+                        continue;
+
+                if (isempty(path_startswith(path, prefix)))
+                        continue;
+
+                p = strdup(path);
+                if (!p) {
+                        r = log_oom_debug();
+                        goto fail;
+                }
+
+                mount_fd = open_tree(AT_FDCWD, path, OPEN_TREE_CLONE | OPEN_TREE_CLOEXEC | AT_RECURSIVE);
+                if (mount_fd < 0) {
+                        r = log_debug_errno(errno, "Failed to open tree of mounted filesystem '%s': %m", path);
+                        goto fail;
+                }
+
+                if (!GREEDY_REALLOC(mounts, n + 1)) {
+                        r = log_oom_debug();
+                        goto fail;
+                }
+
+                mounts[n++] = (SubMount) {
+                        .path = TAKE_PTR(p),
+                        .mount_fd = TAKE_FD(mount_fd),
+                };
+        }
+
+        typesafe_qsort(mounts, n, sub_mount_compare);
+        sub_mount_drop(mounts, n);
+
+        *ret_mounts = TAKE_PTR(mounts);
+        *ret_n_mounts = n;
+        return 0;
+
+fail:
+        sub_mount_free_n(mounts, n);
+        return r;
+}
+
+static int mount_sub_mounts(SubMount *mounts, size_t n) {
+        assert(mounts || n == 0);
+
+        for (size_t i = 0; i < n; i++) {
+                if (!mounts[i].path || mounts[i].mount_fd < 0)
+                        continue;
+
+                (void) mkdir_p_label(mounts[i].path, 0755);
+
+                /* And place the cloned version in its place */
+                if (move_mount(mounts[i].mount_fd, "", AT_FDCWD, mounts[i].path, MOVE_MOUNT_F_EMPTY_PATH) < 0)
+                        return log_debug_errno(errno, "Failed to move mount_fd to '%s': %m", mounts[i].path);
+        }
+
+        return 0;
+}
+
+int remount_sysfs(const char *path) {
+        SubMount *mounts;
+        size_t n;
+        int r;
+
+        assert(path);
```

Comment:
Maybe add a quick comment what this actually does, i.e. why we need something so involved here. It's not obvious to the random reader I think.